File: src/camping/random-encounters.ts
import {getStringSetting} from '../settings';
import {findRollTableUuidWithFallback, findWorldTableUuid, rollRollTable} from '../roll-tables';
import {DegreeOfSuccess, determineDegreeOfSuccess, StringDegreeOfSuccess} from '../degree-of-success';
import {regions, toOfficialKingmakerRollTableName} from './regions';
import {CampingActivityData} from './activities';
import {Camping, getCampingActivityData} from './camping';
import {getWorldTime, isDayOrNight} from '../time/calculation';
import {isKingmakerInstalled} from '../utils';


async function succeededEncounterFlatCheck(region: string, dc: EncounterDc): Promise<boolean> {
    const flavor = `Rolling Random Encounter for terrain ${region} with Flat DC ${dc.total}`;
    const dieRoll = await new Roll('1d20').evaluate();
    await dieRoll.toMessage({flavor}, {rollMode: 'gmroll'});
    const degreeOfSuccess = determineDegreeOfSuccess(dieRoll.total, dieRoll.total, dc.total);
    return degreeOfSuccess === DegreeOfSuccess.CRITICAL_SUCCESS || degreeOfSuccess === DegreeOfSuccess.SUCCESS;
}

async function findCreatureRollTable(game: Game, region: string): Promise<string> {
    const key = isKingmakerInstalled(game) ? toOfficialKingmakerRollTableName(region, regions.get(region)!) : region;
    return (await findRollTableUuidWithFallback(game, key, 'pf2e-kingmaker-tools.kingmaker-tools-random-encounters'))!;
}

export async function rollRandomEncounter(game: Game, region: string, dc: EncounterDc, forgoFlatCheck = false): Promise<boolean> {
    const creatureRollTable = await findCreatureRollTable(game, region);
    const worldEncounterTable = getStringSetting(game, 'proxyEncounterTable')?.trim();
    const encounterTypeRollTable = findWorldTableUuid(game, worldEncounterTable)!;
    const rollMode = getStringSetting(game, 'randomEncounterRollMode') as unknown as keyof CONFIG.Dice.RollModes;
    const rollOptions: Partial<RollTable.DrawOptions> = {rollMode};

    if (forgoFlatCheck || await succeededEncounterFlatCheck(region, dc)) {
        if (encounterTypeRollTable === undefined ||
            encounterTypeRollTable === null ||
            encounterTypeRollTable.trim() === '') {
            console.log('Rolling on creature table ' + creatureRollTable);
            await rollRollTable(game, creatureRollTable, rollOptions);
        } else {
            const typeResult = await rollRollTable(game, encounterTypeRollTable, rollOptions);
            /* eslint-disable @typescript-eslint/no-explicit-any */
            const tableResult = typeResult.draw.results[0] as any;
            if (tableResult?.text?.trim() === 'Creature') {
                console.log('Rolling on creature table ' + creatureRollTable);
                await rollRollTable(game, creatureRollTable, rollOptions);
            }
        }
        return true;
    }
    return false;
}


export interface ActivityModifier {
    day: number,
    night: number
}

export function getActivityModifier(
    campingActivityData: CampingActivityData,
    result: StringDegreeOfSuccess | null,
): ActivityModifier {
    if (campingActivityData.modifyRandomEncounterDc) {
        return {
            day: campingActivityData.modifyRandomEncounterDc.day,
            night: campingActivityData.modifyRandomEncounterDc.night,
        };
    } else if (result !== null) {
        const mods = campingActivityData[result]?.modifyRandomEncounterDc;
        return {
            day: mods?.day ?? 0,
            night: mods?.night ?? 0,
        };
    } else {
        return {day: 0, night: 0};
    }
}

export interface EncounterDc {
    modifier: number;
    activityModifier: number;
    dc: number;
    total: number;
}

export function getEncounterDC(data: Camping, game: Game): EncounterDc {
    const currentRegionData = regions.get(data.currentRegion);
    const modifier = data.encounterModifier;
    const activityModifier = getCampingActivityData(data)
        .map(campingActivityData => {
            const campingActivity = data.campingActivities
                .find(ca => ca.activity === campingActivityData.name);
            const result = campingActivity?.result ?? null;
            const actorUuid = campingActivity?.actorUuid ?? null;
            if (actorUuid !== null && (result !== null || campingActivityData.skills.length === 0)) {
                const mods = getActivityModifier(campingActivityData, result);
                return mods[isDayOrNight(getWorldTime(game))];
            } else {
                return 0;
            }
        }).reduce((a, b) => a + b, 0);
    const dc = currentRegionData?.encounterDC ?? 0;
    return {
        modifier,
        activityModifier,
        dc,
        total: modifier + dc + activityModifier,
    };
}

File: src/jsMain/kotlin/at/posselt/pfrpg2e/camping/RandomEncounters.kt
package at.posselt.pfrpg2e.camping

import at.posselt.pfrpg2e.actor.party
import at.posselt.pfrpg2e.camping.dialogs.RegionSetting
import at.posselt.pfrpg2e.data.checks.DegreeOfSuccess
import at.posselt.pfrpg2e.data.checks.RollMode
import at.posselt.pfrpg2e.fromCamelCase
import at.posselt.pfrpg2e.utils.*
import com.foundryvtt.core.Game
import com.foundryvtt.core.documents.RollTable
import com.foundryvtt.core.ui
import com.foundryvtt.pf2e.actor.PF2ENpc

suspend fun rollRandomEncounter(
    game: Game,
    actor: PF2ENpc,
    includeFlatCheck: Boolean
): Boolean {
    actor.getCamping()?.let { camping ->
        val currentRegion = camping.findCurrentRegion() ?: camping.regionSettings.regions.firstOrNull()
        currentRegion?.let { region ->
            val partyLevel = game.party()?.level ?: 1
            return rollRandomEncounter(
                camping = camping,
                includeFlatCheck = includeFlatCheck,
                region = region,
                isDay = game.getPF2EWorldTime().time.isDay(),
                partyLevel = partyLevel,
            )
        }
    }
    return false
}

private suspend fun rollRandomEncounter(
    camping: CampingData,
    includeFlatCheck: Boolean,
    region: RegionSetting,
    isDay: Boolean,
    partyLevel: Int,
): Boolean {
    val table = region.rollTableUuid?.let { fromUuidTypeSafe<RollTable>(it) }
    if (table == null) {
        if (region.rollTableUuid != null) {
            ui.notifications.error("Could not find random encounter roll table for region ${region.name}")
        }
        return false
    }
    val rollMode = fromCamelCase<RollMode>(camping.randomEncounterRollMode) ?: RollMode.GMROLL
    val proxyTable = camping.proxyRandomEncounterTableUuid?.let { fromUuidTypeSafe<RollTable>(it) }
    val dc = findEncounterDcModifier(camping, isDay)
    val rollCheck = if (includeFlatCheck) {
        d20Check(
            dc = dc,
            flavor = "Rolling Random Encounter for terrain ${region.name} with Flat DC $dc",
            rollMode = rollMode,
        ).degreeOfSuccess.succeeded()
    } else {
        true
    }
    if (rollCheck) {
        val proxyResult = proxyTable?.rollWithDraw(rollMode = rollMode)
            ?.draw
            ?.results
            ?.get(0)
            ?.text
            ?.trim()
            ?: "Creature"
        if (proxyResult == "Creature") {
            table.rollWithDraw(rollMode = rollMode)
        }
        if (camping.campingActivities.any {
                val result = it.parseResult()
                it.isPrepareCampsite() && result != null && result != DegreeOfSuccess.CRITICAL_FAILURE
            }) {
            postCombatEffects(
                activeActivities = camping.alwaysPerformActivities.toSet() +
                        camping.campingActivities
                            .filter { it.actorUuid != null }
                            .map { it.activity },
                partyLevel = partyLevel
            )
        }
        return true
    }
    return false
}

fun findEncounterDcModifier(
    camping: CampingData,
    isDay: Boolean
): Int = (camping.findCurrentRegion()?.encounterDc ?: 0) +
        calculateModifierIncrease(camping, isDay) +
        camping.encounterModifier

private fun calculateModifierIncrease(camping: CampingData, isDay: Boolean): Int =
    camping.groupActivities().asSequence()
        .filter { it.done() || camping.alwaysPerformActivities.contains(it.data.name) }
        .map { (data, activity) -> calculateModifierIncrease(data, isDay, activity.parseResult()) }
        .sum()


private fun calculateModifierIncrease(
    data: CampingActivityData,
    isDay: Boolean,
    checkResult: DegreeOfSuccess?
): Int {
    val activityMod = data.modifyRandomEncounterDc?.atTime(isDay) ?: 0
    val resultMod = checkResult?.let { result ->
        data.getOutcome(result)
            ?.modifyRandomEncounterDc
            ?.atTime(isDay)
    } ?: 0
    return activityMod + resultMod
}
